bundler/lib/bundler/source/rubygems.rb

```diff
@@ -380,7 +380,7 @@ def installed_specs
 
       def cached_specs
         @cached_specs ||= begin
-          idx = @allow_local ? installed_specs.dup : Index.new
+          idx = Index.new
```

Comment:
is this a change in behavior here?

Reply:
When `@allow_local` is true, [`installed_specs` is always merged on top of `cached_specs`](https://github.com/rubygems/rubygems/pull/6931/files#diff-cae021dce97344756b9347e262ca14ff32d50fdc3b48930398f37a68515afe2eR135), replacing every installed_spec from this duped version with the exact same spec from installed_specs. Before this change it was like the following pseudo code:
```
# @allow_cached = true && allow_local = true
  specs = installed.merge(cached).merge(installed)
# @allow_remote = true && @allow_local = false
  specs = cached
```
The symptom I was tracing was "why do we run into so many duplicate specs? Why is there a list of specs in (formerly) `@all_specs` that contains multiple copies of the same spec and source?" It's because this was causing a guaranteed dup of every installed spec every time allow_local was enabled.

Reply:
are there any places that are using `installed_specs` independently? because I think it's possible for a spec to be cached but not installed, and iirc those are public methods on `Source::Rubygems`? It's possible (maybe event likely) that the answer is no, I'm just worried about something not being caught by the specs that someone might be relying on

Reply:
`cached_specs` is only called once by `specs` in the way I described. It should be private imho.

Reply:
To put it another way, you could write `specs` like so without this fix:

```ruby
      def specs
        @specs ||= begin
          index = @allow_remote ? remote_specs.dup : Index.new
          index.merge!(installed_specs) if @allow_local
          index.merge!(cached_specs) if @allow_cached || @allow_remote
          index.merge!(installed_specs) if @allow_local
          index
        end
```

Since `cached_specs` is never used outside of this context, the only thing that matters is it's result for this `specs`.

Reply:
Sweet, thanks for diving in & verifying!

Reply:
Oh, I was wrong, `cached_specs` is already `protected`, as is `installed_specs`. They aren't public.